packages/signature-controller/src/SignatureController.ts

```diff
@@ -492,7 +492,11 @@ export class SignatureController extends BaseControllerV2<
       const cleanMessageParams = await messageManager.approveMessage(msgParams);
       const signature = await getSignature(cleanMessageParams);
 
-      messageManager.setMessageStatusSigned(messageId, signature);
+      this.hub.emit(`${methodName}:signed`, { signature, messageId });
+
+      if (!cleanMessageParams.deferSetAsSigned) {
```

Comment:
> However the dApp will still be waiting for a promise that will only resolve when this status is changed by the custody logic in the extension.

> Is the typical use case then that a dApp will wait for minutes (hours?) to receive the signature via the promise, or do they poll the message state somehow?

That's true. Dapps await a promise from calls to the message signing RPC commands. I made an animation below that roughly shows the user experience seen today.

It is correct that the popup closes with a pending promise - in MMI we also show the user additional instructions.

There's no way for a dapp to poll for signed messages that we're aware of. However, we've found that all the dapps our users use are OK with awaiting indefinitely.

![signed_message-optimised](https://github.com/MetaMask/core/assets/1504499/744b99e5-c9c9-46af-b37c-5eb468073bdf)

> None of this state is persisted so that would only be possible until the extension is restarted or the browser is closed.

That's correct - although the dapp would lose the connection to the ethereum provider after a restart anyway.